### ISSUE TITLE
jQuery.imagesLoaded does not work with FireFox

### ISSUE THREAD

**Issue by justnorris** [desandro/imagesloaded#191]
![screen shot 2015-04-30 at 12 11 07 pm](https://cloud.githubusercontent.com/assets/988095/7409911/ee3c3128-ef32-11e4-979b-42e34bca23d6.png)

*Note: I didn't just "quickly take a screenshot". I waited for about a minute or so (in the meantime I moved the "about windows" in view). The images are never loaded*

*Note 2: In the screenshot it looks like no images are loaded at all. That's not true either. I just moved the "about windows" on top of the loaded images accidentally.*

On not that good networks ( that's what the network link conditioner is there for ) imagesLoaded just stops trying to load the images at all. No 'fail', no 'always', it just stops and does nothing.

I used this for the screenshot: http://codepen.io/desandro/pen/bIFyl 

imagesLoaded vanilla version works just fine with firefox and problematic network. For some reason the jQuery version is what causes problems.

**Comment by desandro** [comment]
Thanks for reporting this bug. I'm not able to reproduce this bug in Firefox v38. Sounds like a networking problem. Can you clarify that this problem only happens:

+ with poor networking
+ in Firefox
+ with jQuery

**Comment by justnorris** [comment]
1. Yes
2. As far as I've tested - Yes.
3. Yes.


Did you try with "Network Link Conditioner" ? It only happens to me with that thing turned on. Stumbled on this because users reported the site not loading properly ( I'm using imagesLoaded to fade the site in ).

**Comment by justnorris** [comment]
FireFox 38b9
![screen shot 2015-05-03 at 9 45 39 pm](https://cloud.githubusercontent.com/assets/988095/7446339/c84f33ae-f1dd-11e4-9d91-849fa2c6949b.png)

It doesn't always freeze. To reproduce click "add images" about 3 times, then wait for half of them to load, then click twice again. If that doesn't work - clear cache and try again.

It's a rather random error, but when many images are involved - it has a high risk of hanging in FireFox.

**Comment by justnorris** [comment]
Update: Found that it's broken on vanilla version too.
![screen shot 2015-05-03 at 9 50 05 pm](https://cloud.githubusercontent.com/assets/988095/7446354/753fb278-f1de-11e4-8c83-469fc0936389.png)

**Comment by justnorris** [comment]
Tested Safari 8 with same network settings - works. Pushed Chrome to the limit with 44% packet drop and 800ms DNS delay - it took about 10 minutes to load, but everything worked out fine. 

It's a FireFox Thing.


Another hint of what might the issue be - when it's stuck - if I "add more images", the images that were stuck previously are "unstuck" and everything loads properly.

**Comment by justnorris** [comment]
**beep beep**
Any updates here ? It's kind of a big issue for me.

**Comment by desandro** [comment]
Sorry, I won't be able to dig into this one. It's a deep edge case. I'll label this as _help wanted_.

**Comment by justnorris** [comment]
In my mind - 7% of browser market-share isn't a deep edge case, considering that Flickity, Masonry, Packery are all based on imagesLoaded. It's a critical instability issue at the moment. 

I'd help, but I have no idea even where to begin. Any ideas on what might be wrong with FireFox ?

**Comment by justnorris** [comment]
Here is a fully loaded "webster hall" homepage, found in your masonry demo links- with the same issue:
![screen shot 2015-05-08 at 3 58 37 pm](https://cloud.githubusercontent.com/assets/988095/7536768/3f34cfea-f59b-11e4-9a41-8c17a135ca20.png)

**Comment by crazoter** [comment]
I'm currently on Firefox 37.0.2 & I'm also experiencing the issue. I don't think the issue is isolated to only slow networks (although it does help reproduce the error) because I managed to reproduce the error consistently whilst having a download speed of 3.77 Mbps.

I'm no expert so I'm took a rather thorough (and possibly quite difficult to read) approach to this issue. The details can be found below.

tl;dr; in the Resource.prototype.check function, I changed 
```javascript
var proxyImage = new Image();
```
to 
```javascript
var proxyImage = this.img;
```
where this.img is the actual Image element we are trying to track. I simply passed it into the Resource constructor when it was created here and modified the constructor to store the variable.

```javascript
//modified constructor
var resource = cache[ this.img.src ] || new Resource( this.img.src,this.img );
```
```javascript
//modified constructor
function Resource( src,img ) {
    this.src = src;
    this.img = img;
    // add to cache
    cache[ src ] = this;
  }
```

<b>Limitations of this fix:</b>
I'm not very sure about this - this technique may not work when attempting to detect reloading images e.g. if an image is loaded using the Image element that has already previously loaded an image. I have not tested this to see if it will still work then.

This fixed the issue on firefox. I did not test extensively on other browsers, but chrome still worked fine.

<b>Details:</b>
First, I have to understand the process, so I will state my understanding:
<b>How I understand imagesLoaded works:</b>
1. an imagesLoaded object is created with the dom elements to be tracked (or a query selector string which is later used to obtain the elements).
* getImages() is called, which is basically getting all image elements to be tracked.
2. imagesLoaded.prototype.check is called (with a slight time delay, 4-10ms) which basically involves looping through every image, checking if they are complete and binding the complete eventhandler on them.
* LoadingImage.prototype.check is called for each image (to check if they are complete).
   * First, it checks if the image was loaded before (check if cached). If not, it initializes a new Resource object for that image.
   * Usually that returns false, so it then uses the image.complete and naturalwidth to check if it is complete manually. This is form of checking is only performed once, 4-10ms after the imagesLoaded object is created.
   * Unless your image was already loaded when the imagesLoaded object was created, that will return false and it will call Resource.prototype.check which contains the image's src.
       * At this point, resource determines if the image is loaded by simulation - it creates a new Image, binds the 'load' and 'error' to it using eventie and sets the src so that it begins loading.

<b>Observations:</b>
  * I don't think the issue lies in LoadingImage.prototype.check because all images are correctly obtained and checked here.
  * I don't think the issue lies in the Firefox Images because I had looped through all the images stored in this.Images after I was certain they were loaded and all of them had true on their complete property and a non-undefined naturalWidth property. Of course, because these properties are only checked at the start, their use is limited in this scenario.
  * However, when an image "fails to load" the event is not caught in either 'load' or 'error', so the error must be related to the binding of the event to detect if the image is loaded / error. This is the method I speak of:
```javascript
    Resource.prototype.check = function() {
    // only trigger checking once
    if ( this.isChecked ) {
      return;
    }
    // simulate loading on detached element
    var proxyImage = new Image();
    eventie.bind( proxyImage, 'load', this );
    eventie.bind( proxyImage, 'error', this );
    proxyImage.src = this.src;
    // set flag
    this.isChecked = true;
  };
```
  * I don't think the issue lies in eventie because eventie just binds the event. However, I think the problem lies in creating a new Image() and using it to check if the image is loaded - in Firefox it apparently doesn't succeed all the time for some reason. When I passed the image element into the Resource constructor and did 
```javascript
 var proxyImage = this.img;//this.img is the actual dom Image we are trying to track
```
it worked 5/5 times (100% success rate) I tried (as opposed to the 1/5 (20% success rate) with creating a new Image).

**Comment by duynguyen1990** [comment]
Hi,

I have a trouble with imagesloaded js, when i try with ctrl + f5 the browser, the issue sometimes happens. I show the console in firebug and notice that one image is checked by false ( "check" event return false), and that image has 744.00KB and load time 7.15s. Below is images

P/s sorry for bad english
Wait for your advice!
![imageloaded1](https://cloud.githubusercontent.com/assets/11290552/7855187/da134f38-0549-11e5-847d-8a974f99268b.png)
![imageloaded2](https://cloud.githubusercontent.com/assets/11290552/7855191/ddc6736c-0549-11e5-8d5a-cbf30cb351f6.png)
![imageloaded3](https://cloud.githubusercontent.com/assets/11290552/7855193/e362ed00-0549-11e5-8396-f93053b554bf.png)

**Comment by crazoter** [comment]
Did you use the imagesloaded.pkgd.js which includes the commit pushed by dmoreno? It should fix the issue. 
You can get the file from this branch:
https://github.com/hi-impact/imagesloaded/tree/fix-issue-191

**Comment by tgmweb** [comment]
@crazoter - I had the same issue with Firefox and your fix worked for me. 

Thanks!

BTW - this issue is not limited to slow networks. Our images are hosted on s3/cloudfront and we're in NYC with high speed FIOS.

Tom.

**Comment by tmslnz** [comment]
@crazoter Thank you for this; it fixed my issue as well.
@desandro Wonder why this fix isn’t merged in yet (or manually integrated)?

**Comment by desandro** [comment]
@crazoter Thanks for looking deeper into this! This is tremendous work. I'll be incorporating your insights in the next refactor.

**Comment by vejersele** [comment]
@desandro you probably also want to try the following:

```javascript
Resource.prototype.check = function() {
    // only trigger checking once
    if ( this.isChecked ) {
      return;
    }
    // simulate loading on detached element
    this.proxyImage = new Image();
    eventie.bind( this.proxyImage, 'load', this );
    eventie.bind( this.proxyImage, 'error', this );
    this.proxyImage.src = this.src;
    // set flag
    this.isChecked = true;
  };
```

It could be that the proxyImage variable is getting garbage collected because it's defined locally. Assigning proxyImage as a class property will prevent this from happening ...

**Comment by m4olivei** [comment]
We're also seeing this issue, and the proposed fix from @crazoter and @dmoreno seems to fix the problem.  Can't say I understand why that patch fixes things, but seems to.  Testing ongoing.  Very interested in seeing this make it into a future release.  We're relying on imagesLoaded to "reveal" parts of our website.

**Comment by tparnell8** [comment]
:+1:  for this fix What can we do to get this in?

**Comment by justnorris** [comment]
@desandro I thought you [quit twitter](http://metafizzy.co/blog/full-time-fizzy/) to work full-time on your plugins. This seems like a pretty critical issue, isn't it? It's in the core of many of your other libraries, I don't understand why is this issue ignored. 

<img width="779" alt="screen shot 2015-10-27 at 09 57 36" src="https://cloud.githubusercontent.com/assets/988095/10754713/5eb52f9a-7c91-11e5-9a10-65534ad3520b.png">

**Comment by desandro** [comment]
:tada:  imagesLoaded v3.2.0 integrates @crazoter's [fix for Firefox](https://github.com/desandro/imagesloaded/blob/v3.2.0/imagesloaded.js#L287-L289) :tada:  [Try it out](http://imagesloaded.desandro.com/#install) and report back if this issue has been resolved or not :tangerine:

**Comment by gsmeira** [comment]
Just in time! :D

It's working for me here in firefox, and now the plugin is detecting reloaded images. I'm using another plugin, ResponseJS, and when the src change, the imagesloaded is detecting that change.

Thanks @desandro for the plugin and thanks @crazoter for the fix!! :+1:

**Comment by desandro** [comment]
I'm closing this issue as resolved. If you run into an issue with imagesLoaded or Firefox, please open a new issue.

**Comment by lucacattide** [comment]
On FireFox 61.0.2 issue back again
